HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+        self.vFunc_Sha_Adj = vFunc_Sha_Adj
+        self.ShareFunc_Adj = ShareFunc_Adj
+        self.dvdmFunc_Sha_Adj = dvdmFunc_Sha_Adj
+        self.dvdnFunc_Sha_Adj = dvdnFunc_Sha_Adj
+
+        self.vFunc_Sha_Fxd = vFunc_Sha_Fxd
+        self.ShareFunc_Fxd = ShareFunc_Fxd
+        self.dvdmFunc_Sha_Fxd = dvdmFunc_Sha_Fxd
+        self.dvdnFunc_Sha_Fxd = dvdnFunc_Sha_Fxd
+        self.dvdsFunc_Sha_Fxd = dvdsFunc_Sha_Fxd
+
+
+# Class for the consumption stage solution
+class RiskyContribCnsSolution(MetricObject):
+    """
+    A class for representing the solution to the consumption stage of the
+    'RiskyContrib' model.
+    
+    Parameters
+    ----------
+    vFunc_Cns : ValueFunc3D
+        Stage-value function over normalized liquid resources, normalized
+        iliquid resources, and income contribution share.
+    cFunc : Interp3D
+        Consumption function over normalized liquid resources, normalized
+        iliquid resources, and income contribution share.
+    dvdmFunc_Cns : MargValueFunc3D
+        Marginal value function over normalized liquid resources.
+    dvdnFunc_Cns : MargValueFunc3D
+        Marginal value function over normalized iliquid resources.
+    dvdsFunc_Cns : Interp3D
+        Marginal value function over income contribution share.
+    """
+
+    distance_criteria = ["dvdmFunc_Cns", "dvdnFunc_Cns"]
+
+    def __init__(
+        self,
+        # Consumption stage
+        vFunc_Cns=None,
+        cFunc=None,
+        dvdmFunc_Cns=None,
+        dvdnFunc_Cns=None,
+        dvdsFunc_Cns=None,
+    ):
+
+        if vFunc_Cns is None:
+            vFunc_Cns = NullFunc()
+        if cFunc is None:
+            cFunc = NullFunc()
+        if dvdmFunc_Cns is None:
+            dvdmFunc_Cns = NullFunc()
+        if dvdnFunc_Cns is None:
+            dvdmFunc_Cns = NullFunc()
+        if dvdsFunc_Cns is None:
+            dvdsFunc_Cns = NullFunc()
+
+        self.vFunc_Cns = vFunc_Cns
+        self.cFunc = cFunc
+        self.dvdmFunc_Cns = dvdmFunc_Cns
+        self.dvdnFunc_Cns = dvdnFunc_Cns
+        self.dvdsFunc_Cns = dvdsFunc_Cns
+
+
+# Class for the solution of a whole period
+class RiskyContribSolution(MetricObject):
+    """
+    A class for representing the solution to a full time-period of the
+    'RiskyContrib' agent type's problem.
+    
+    Parameters
+    ----------
+    Reb : RiskyContribRebSolution
+        Solution to the period's rebalancing stage.
+    Sha : RiskyContribShaSolution
+        Solution to the period's contribution-share stage.
+    Cns : RiskyContribCnsSolution
+        Solution to the period's consumption stage.
+    """
+
+    # Solutions are to be compared on the basis of their sub-period solutions
+    distance_criteria = ["stage_sols"]
+
+    def __init__(self, Reb, Sha, Cns):
+
+        # Dictionary of stage solutions
+        self.stage_sols = {"Reb": Reb, "Sha": Sha, "Cns": Cns}
+
+
+# %% Auxiliary functions and transition equations for the RiskyContrib model.
```

Comment:
I like these. Elegant.
I wonder if there's a better place to put them than at the end of the file like this?
not a blocker at all, just a style/design puzzle.

Reply:
I initially thought that, as intrinsic properties of the model, they ought to be inside the agent-class. However, they are used by solvers who do not and should not deal with the "whole agent" but merely a period, so they would have no way to access them.

Open to suggestions, however.